### PR TITLE
nodefs: add TestReadDirStress

### PR DESCRIPTION
**Pull request by rfjakob**
This currently fails on Linux 5.0 and may be related to
https://github.com/hanwen/go-fuse/issues/287 .

```
1 jakob@brikett:~/go/src/github.com/hanwen/go-fuse/nodefs$ go test
21:42:47.356529 writer: Write/Writev failed, err: 2=no such file or directory. opcode: RELEASE
21:42:47.598309 writer: Write/Writev failed, err: 22=invalid argument. opcode: READDIRPLUS
21:42:47.604424 writer: Write/Writev failed, err: 22=invalid argument. opcode: READDIRPLUS
21:42:47.606073 writer: Write/Writev failed, err: 22=invalid argument. opcode: READDIRPLUS
--- FAIL: TestReadDirStress (0.36s)
    simple_test.go:270: goroutine 2 iteration 5: readdirent: input/output error
    simple_test.go:270: goroutine 1 iteration 9: readdirent: input/output error
    simple_test.go:270: goroutine 3 iteration 10: readdirent: input/output error
    simple_test.go:43: /usr/bin/fusermount: entry for /tmp/TestReadDirStress639795934/mnt not found in /etc/mtab
         (code exit status 1)

FAIL
exit status 1
FAIL	github.com/hanwen/go-fuse/nodefs	0.994s
```

### CODE REVIEW

**Comment by hanwen** [comment]
this is extremely weird: this is broken over here (linux 5.0.5-200.fc29.x86_64), leading to hangs with the following trace:

goroutine 4 [runnable]:
syscall.Syscall(0xd9, 0x6, 0xc00029d000, 0x1000, 0xffffffffffffffff, 0x0, 0x6b)
	/usr/lib/golang/src/syscall/asm_linux_amd64.s:18 +0x5
syscall.Getdents(0x6, 0xc00029d000, 0x1000, 0x1000, 0xc00029dfc3, 0x28, 0xc0002e2cc0)
	/usr/lib/golang/src/syscall/zsyscall_linux_amd64.go:465 +0x5a
github.com/hanwen/go-fuse/nodefs.(*loopbackDirStream).load(0xc000069780, 0xc00029dfc3)
	/home/hanwen/vc/go-fuse/nodefs/dirstream_linux.go:72 +0x58
github.com/hanwen/go-fuse/nodefs.(*loopbackDirStream).Next(0xc000069780, 0xc00029c001, 0xc0002ec630, 0x28, 0xc0001cc8c0, 0x0)
	/home/hanwen/vc/go-fuse/nodefs/dirstream_linux.go:64 +0x11a
github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus(0xc000110000, 0xc00014e000, 0xc000150198, 0xc000010f60, 0x1000)
	/home/hanwen/vc/go-fuse/nodefs/bridge.go:844 +0x474
github.com/hanwen/go-fuse/fuse.doReadDirPlus(0xc0000fc140, 0xc000150000)
	/home/hanwen/vc/go-fuse/fuse/opcode.go:162 +0xf6
github.com/hanwen/go-fuse/fuse.(*Server).handleRequest(0xc0000fc140, 0xc000150000, 0xc000000000)
	/home/hanwen/vc/go-fuse/fuse/server.go:469 +0x294
github.com/hanwen/go-fuse/fuse.(*Server).loop(0xc0000fc140, 0xc00000c001)
	/home/hanwen/vc/go-fuse/fuse/server.go:436 +0x18f
created by github.com/hanwen/go-fuse/fuse.(*Server).readRequest
	/home/hanwen/vc/go-fuse/fuse/server.go:303 +0x3a1

ie. the kernel is hung while doing a readdir on a dir under /tmp

File: nodefs/simple_test.go [hunk 1]
```diff
@@ -250,6 +250,38 @@ func TestReadDir(t *testing.T) {
...
+		for i := 1; i < 100; i++ {
+			f, err := os.Open(tc.mntDir)
+			if err != nil {
+				t.Error(err)
+				return
+			}
+			_, err = f.Readdirnames(-1)
+			if err != nil {
+				t.Errorf("goroutine %d iteration %d: %v", gr, i, err)
+				return
```
Comment (hanwen) [inline nodefs/simple_test.go]: nit, this leaks the fd, leading to error on unmount.

Comment (rfjakob) [reply nodefs/simple_test.go]: Uh, right

**Comment by hanwen** [comment]
(the hang I saw happens when I _remove_ parallel dirops support.)

File: nodefs/simple_test.go (hunk 1 above)
Comment (rfjakob) [reply nodefs/simple_test.go]: Fixed

**Comment by hanwen** [comment]
removing cap_readdirplus ,  cap_parallel_ops , I see

normal readdir: 

22:20:19.669489 rx 1366: OPENDIR i1 
22:20:19.669699 tx 1366:     OK, {Fh 1 }
22:20:19.669755 rx 1374: READDIR i1 {Fh 1 [0 +4096)  L 0 RDONLY,0x8000} 
22:20:19.669878 tx 1374:     OK,  4096b data "\xf2\xa5\x1b\x00\x00\x00\x00\x00"...
22:20:19.669959 rx 1376: READDIR i1 {Fh 1 [65 +4096)  L 0 RDONLY,0x8000} 
22:20:19.669978 tx 1376:     OK,  3008b data "\xf0\x9e\x1b\x00\x00\x00\x00\x00"...
22:20:19.670056 rx 1378: READDIR i1 {Fh 1 [112 +4096)  L 0 RDONLY,0x8000} 
22:20:19.670076 tx 1378:     OK

..

22:20:19.671017 rx 1404: OPENDIR i1 
22:20:19.684388 tx 1404:     OK, {Fh 1 }
22:20:19.684447 rx 1406: OPENDIR i1 
22:20:19.684476 tx 1406:     OK, {Fh 2 }
22:20:19.684569 rx 1408: READDIR i1 {Fh 1 [0 +4096)  L 0 RDONLY,0x8000} 
22:20:19.670190 rx 1384: OPENDIR i1 
22:20:19.684773 tx 1384:     OK, {Fh 3 }
22:20:19.684938 tx 1408:     OK,  4096b data "\xf2\xa5\x1b\x00\x00\x00\x00\x00"...
22:20:19.685085 rx 1410: READDIR i1 {Fh 2 [0 +4096)  L 0 RDONLY,0x8000} 
22:20:19.685254 tx 1410:     OK,  4096b data "\xf2\xa5\x1b\x00\x00\x00\x00\x00"...
22:20:19.685350 rx 1412: READDIR i1 {Fh 3 [0 +4096)  L 0 RDONLY,0x8000} 
22:20:19.685436 rx 1414: RELEASEDIR i1 {Fh 2 0x8000  L0} 
22:20:19.685439 gorouting 1 it 3
22:20:19.685490 tx 1412:     OK,  4096b data "\xf2\xa5\x1b\x00\x00\x00\x00\x00"...
22:20:19.685611 rx 1418: READDIR i1 {Fh 1 [65 +4096)  L 0 RDONLY,0x8000} 
22:20:19.685653 rx 1416: OPENDIR i1 
22:20:19.685704 tx 1416:     OK, {Fh 2 }
22:20:19.685474 tx 1414:     OK
22:20:19.685660 tx 1418:     9=bad file descriptor,  2944b data "\xf0\x9e\x1b\x00\x00\x00\x00\x00"...

in rx/tx 1418, the read from pos 65 should return 3008 bytes, ie. 64 bytes are missing (exactly 1 entry), and the error code is EBADF.

**Comment by rfjakob** [comment]
Can confirm on `5.0.4-200.fc29.x86_64`. Removing `CAP_PARALLEL_DIROPS` turns this into a deadlock in the kernel. Oh dear.

```
$ pstree -p 7368
nodefs.test(7368)─┬─cat(7392)
                  ├─{nodefs.test}(7370)
                  └─{nodefs.test}(7395)

$ sudo cat /proc/7368/stack
[<0>] request_wait_answer+0x12b/0x200 [fuse]
[<0>] __fuse_request_send+0x69/0x70 [fuse]
[<0>] fuse_readdir+0x271/0xef1 [fuse]
[<0>] iterate_dir+0x13c/0x190
[<0>] ksys_getdents64+0x9c/0x130
[<0>] __x64_sys_getdents64+0x16/0x20
[<0>] do_syscall_64+0x5b/0x160
[<0>] entry_SYSCALL_64_after_hwframe+0x44/0xa9
[<0>] 0xffffffffffffffff

$ sudo cat /proc/7370/stack
[<0>] __fdget_pos+0x42/0x50
[<0>] ksys_getdents64+0x81/0x130
[<0>] __x64_sys_getdents64+0x16/0x20
[<0>] do_syscall_64+0x5b/0x160
[<0>] entry_SYSCALL_64_after_hwframe+0x44/0xa9
[<0>] 0xffffffffffffffff

$ sudo cat /proc/7395/stack
[<0>] fuse_lock_inode+0x28/0x30 [fuse]
[<0>] fuse_readdir+0x262/0xef1 [fuse]
[<0>] iterate_dir+0x13c/0x190
[<0>] ksys_getdents64+0x9c/0x130
[<0>] __x64_sys_getdents64+0x16/0x20
[<0>] do_syscall_64+0x5b/0x160
[<0>] entry_SYSCALL_64_after_hwframe+0x44/0xa9
[<0>] 0xffffffffffffffff
```

**Comment by hanwen** [comment]
the missing entry is consistent loopbackDirStream failing to load the next bit of dirstream. It looks like somethingn is closing file handles behind our backs.

**Comment by rfjakob** [comment]
Oh, it's not the kernel that is deadlocked, it's just that the kernel is waiting for reply from us:
```
[<0>] request_wait_answer+0x12b/0x200 [fuse]
```
That's consistent with two outstanding requests:
```
$ cat /sys/fs/fuse/connections/91/waiting 
2
```

**Comment by hanwen** [comment]
but we're using loopback to acces a directory under /tmp/. Until we get dents from /tmp/, we can't return to the kernel.

**Comment by rfjakob** [comment]
Hmm, I see your point.

OTOH, this seems to fix all problems:

```diff
$ git diff
diff --git a/nodefs/dirstream_linux.go b/nodefs/dirstream_linux.go
index 42006fe..adf7ea1 100644
--- a/nodefs/dirstream_linux.go
+++ b/nodefs/dirstream_linux.go
@@ -38,6 +38,7 @@ func NewLoopbackDirStream(name string) (DirStream, syscall.Errno) {
 
 func (ds *loopbackDirStream) Close() {
        syscall.Close(ds.fd)
+       ds.fd = -1
 }
 
 func (ds *loopbackDirStream) HasNext() bool {
```

But what this means is that Close() gets called while other operations are running - race detector confirms:

```
$ go test -run TestReadDirStress -race
==================
WARNING: DATA RACE
Read at 0x00c00049e058 by goroutine 15:
  github.com/hanwen/go-fuse/nodefs.(*loopbackDirStream).HasNext()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/dirstream_linux.go:45 +0x3e
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:837 +0x1f9
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de

Previous write at 0x00c00049e058 by goroutine 32:
  github.com/hanwen/go-fuse/nodefs.(*loopbackDirStream).Next()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/dirstream_linux.go:52 +0x19b
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:844 +0x847
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de

Goroutine 15 (running) created at:
  github.com/hanwen/go-fuse/fuse.(*Server).readRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:303 +0x5f0
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:414 +0x78
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d

Goroutine 32 (running) created at:
  github.com/hanwen/go-fuse/fuse.(*Server).readRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:303 +0x5f0
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:414 +0x78
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d
==================
==================
WARNING: DATA RACE
Read at 0x00c000068088 by goroutine 15:
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:840 +0x221
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de

Previous write at 0x00c000068088 by goroutine 32:
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:854 +0x7b6
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de

Goroutine 15 (running) created at:
  github.com/hanwen/go-fuse/fuse.(*Server).readRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:303 +0x5f0
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:414 +0x78
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d

Goroutine 32 (running) created at:
  github.com/hanwen/go-fuse/fuse.(*Server).readRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:303 +0x5f0
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:414 +0x78
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d
==================
==================
WARNING: DATA RACE
Read at 0x00c000068090 by goroutine 15:
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:841 +0x249
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de

Previous write at 0x00c000068090 by goroutine 32:
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:853 +0x76f
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de

Goroutine 15 (running) created at:
  github.com/hanwen/go-fuse/fuse.(*Server).readRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:303 +0x5f0
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:414 +0x78
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d

Goroutine 32 (running) created at:
  github.com/hanwen/go-fuse/fuse.(*Server).readRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:303 +0x5f0
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:414 +0x78
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d
==================
==================
WARNING: DATA RACE
Read at 0x00c000066518 by goroutine 7:
  github.com/hanwen/go-fuse/nodefs.(*loopbackDirStream).HasNext()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/dirstream_linux.go:45 +0x3e
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:837 +0x1f9
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d

Previous write at 0x00c000066518 by goroutine 9:
  github.com/hanwen/go-fuse/nodefs.(*loopbackDirStream).Next()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/dirstream_linux.go:52 +0x19b
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:844 +0x847
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de

Goroutine 7 (running) created at:
  github.com/hanwen/go-fuse/nodefs.newTestCase()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/simple_test.go:93 +0x911
  github.com/hanwen/go-fuse/nodefs.TestReadDirStress()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/simple_test.go:254 +0x43
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:865 +0x163

Goroutine 9 (running) created at:
  github.com/hanwen/go-fuse/fuse.(*Server).readRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:303 +0x5f0
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:414 +0x78
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d
==================
==================
WARNING: DATA RACE
Read at 0x00c000092508 by goroutine 7:
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:840 +0x221
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d

Previous write at 0x00c000092508 by goroutine 9:
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:854 +0x7b6
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de

Goroutine 7 (running) created at:
  github.com/hanwen/go-fuse/nodefs.newTestCase()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/simple_test.go:93 +0x911
  github.com/hanwen/go-fuse/nodefs.TestReadDirStress()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/simple_test.go:254 +0x43
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:865 +0x163

Goroutine 9 (running) created at:
  github.com/hanwen/go-fuse/fuse.(*Server).readRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:303 +0x5f0
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:414 +0x78
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d
==================
==================
WARNING: DATA RACE
Read at 0x00c000092510 by goroutine 7:
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:841 +0x249
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d

Previous write at 0x00c000092510 by goroutine 9:
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:853 +0x76f
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de

Goroutine 7 (running) created at:
  github.com/hanwen/go-fuse/nodefs.newTestCase()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/simple_test.go:93 +0x911
  github.com/hanwen/go-fuse/nodefs.TestReadDirStress()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/simple_test.go:254 +0x43
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:865 +0x163

Goroutine 9 (running) created at:
  github.com/hanwen/go-fuse/fuse.(*Server).readRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:303 +0x5f0
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:414 +0x78
  github.com/hanwen/go-fuse/fuse.(*Server).Serve()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:359 +0x7d
==================
==================
WARNING: DATA RACE
Write at 0x00c00049d6f8 by goroutine 74:
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).setStream()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:772 +0x282
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:833 +0xee
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de

Previous read at 0x00c00049d6f8 by goroutine 69:
  github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/nodefs/bridge.go:837 +0x1df
  github.com/hanwen/go-fuse/fuse.doReadDirPlus()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/opcode.go:162 +0x1fa
  github.com/hanwen/go-fuse/fuse.(*Server).handleRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:469 +0x4e2
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:436 +0x1de

Goroutine 74 (running) created at:
  github.com/hanwen/go-fuse/fuse.(*Server).readRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:303 +0x5f0
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:414 +0x78

Goroutine 69 (running) created at:
  github.com/hanwen/go-fuse/fuse.(*Server).readRequest()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:303 +0x5f0
  github.com/hanwen/go-fuse/fuse.(*Server).loop()
      /home/jakob/go/src/github.com/hanwen/go-fuse/fuse/server.go:414 +0x78
==================
--- FAIL: TestReadDirStress (1.72s)
    testing.go:809: race detected during execution of test
FAIL
exit status 1
FAIL	github.com/hanwen/go-fuse/nodefs	1.735s
```

**Comment by hanwen** [comment]
The race detector triggers because you are now writing to the fd.

If this fixes things, it suggests there is a double close for the dirstream somewhere in the code.

**Comment by rfjakob** [comment]
I guess the hanging `(*loopbackDirStream).load` does not operate on the file descriptor it thinks it operates on, but on another one where the fd number has since been reused.

**Comment by hanwen** [comment]
there is a double close 


23:02:59.124872 rx 1368: RELEASEDIR i1 {Fh 3 0x8000  L0} 
23:02:59.144074 double close
23:02:59.144147 tx 1368:     OK
23:02:59.124520 gorouting 1 it 1
panic: double close

goroutine 244 [running]:
log.Panicf(0x60ac36, 0xc, 0x0, 0x0, 0x0)
	/usr/lib/golang/src/log/log.go:333 +0xda
github.com/hanwen/go-fuse/nodefs.(*loopbackDirStream).Close(0xc0002c8680)
	/home/hanwen/vc/go-fuse/nodefs/dirstream_linux.go:51 +0x99
github.com/hanwen/go-fuse/nodefs.(*rawBridge).setStream(0xc0004ba0f0, 0xc000248840, 0xc0000baa98, 0xc000308240, 0xc00028d1a0, 0xc0000a3e50)
	/home/hanwen/vc/go-fuse/nodefs/bridge.go:771 +0x13b
github.com/hanwen/go-fuse/nodefs.(*rawBridge).ReadDir(0xc0004ba0f0, 0xc000248840, 0xc0000baa98, 0xc00008ac30, 0x1000)
	/home/hanwen/vc/go-fuse/nodefs/bridge.go:804 +0x78
github.com/hanwen/go-fuse/fuse.doReadDir(0xc0004bc140, 0xc0000ba900)
	/home/hanwen/vc/go-fuse/fuse/opcode.go:152 +0xf6
github.com/hanwen/go-fuse/fuse.(*Server).handleRequest(0xc0004bc140, 0xc0000ba900, 0xc000000000)
	/home/hanwen/vc/go-fuse/fuse/server.go:469 +0x294
github.com/hanwen/go-fuse/fuse.(*Server).loop(0xc0004bc140, 0xc000239901)
	/home/hanwen/vc/go-fuse/fuse/server.go:436 +0x18f
created by github.com/hanwen/go-fuse/fuse.(*Server).readRequest
	/home/hanwen/vc/go-fuse/fuse/server.go:303 +0x3a1

but using a lock is the better approach than trying to get this right.

**Comment by rfjakob** [comment]
The race detector is also unhappy without the fd write - I think we didn't have a concurrent readdir test before.

Yeah I think this needs some locking

**Comment by hanwen** [comment]
good point. I think we need more concurrent tests in general.

**Comment by hanwen** [comment]
https://github.com/hanwen/go-fuse/commit/a856a74e689b12700414e0bef4e2a1ae13e581bd

for the fix.

merged your test in https://github.com/hanwen/go-fuse/commit/623db2fc1986517f2a4216ecd70c947b09642adf

**Comment by rfjakob** [comment]
Can confirm, looks good here now. Thanks!

**Pull request closed without merging**